itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestStatsReplicationScenariosMigration.java

```diff
@@ -0,0 +1,78 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.parse;
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.apache.hadoop.hive.metastore.messaging.json.gzip.GzipJSONMessageEncoder;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.rules.TestName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Tests statistics replication for ACID tables.
+ */
+public class TestStatsReplicationScenariosMigration extends TestStatsReplicationScenarios {
+  @Rule
+  public final TestName testName = new TestName();
+
+  protected static final Logger LOG = LoggerFactory.getLogger(TestReplicationScenarios.class);
+
+  @BeforeClass
+  public static void classLevelSetup() throws Exception {
+    Map<String, String> overrides = new HashMap<>();
+    overrides.put(MetastoreConf.ConfVars.EVENT_MESSAGE_FACTORY.getHiveName(),
+            GzipJSONMessageEncoder.class.getCanonicalName());
+
+    HashMap<String, String> replicaConfigs = new HashMap<String, String>() {{
+      put("hive.support.concurrency", "true");
+      put("hive.txn.manager", "org.apache.hadoop.hive.ql.lockmgr.DbTxnManager");
+      put("hive.metastore.client.capability.check", "false");
+      put("hive.repl.bootstrap.dump.open.txn.timeout", "1s");
+      put("hive.exec.dynamic.partition.mode", "nonstrict");
+      put("hive.strict.checks.bucketing", "false");
+      put("hive.mapred.mode", "nonstrict");
+      put("mapred.input.dir.recursive", "true");
+      put("hive.metastore.disallow.incompatible.col.type.changes", "false");
+      put("hive.strict.managed.tables", "true");
+    }};
+    replicaConfigs.putAll(overrides);
+
+    HashMap<String, String> primaryConfigs = new HashMap<String, String>() {{
+      put("hive.metastore.client.capability.check", "false");
+      put("hive.repl.bootstrap.dump.open.txn.timeout", "1s");
+      put("hive.exec.dynamic.partition.mode", "nonstrict");
+      put("hive.strict.checks.bucketing", "false");
+      put("hive.mapred.mode", "nonstrict");
+      put("mapred.input.dir.recursive", "true");
+      put("hive.metastore.disallow.incompatible.col.type.changes", "false");
+      put("hive.support.concurrency", "false");
+      put("hive.txn.manager", "org.apache.hadoop.hive.ql.lockmgr.DummyTxnManager");
+      put("hive.strict.managed.tables", "false");
+    }};
+    primaryConfigs.putAll(overrides);
+
+    internalBeforeClassSetup(primaryConfigs, replicaConfigs,
```

Comment:
In migration case, we shall validate if stats are associated with correct writeId. I think, in our tests, it should be pointing to last allocated writeId.

Reply:
The write Id associated with the column stats is stored in Table/Partition object. For migration case, this writeId may not exactly represent the writeId of stats if there's an alter table event after column stats update. Such an alter table event would open a new transaction and update writeId in table/partition object.

Reply:
Yeah. That's true.
I have a question on how txn stats work.
If the writeId in the table/partition is less than the latest allocated writeId in the table, will future queries use this stats for query optimization or not? 
If writeId is not latest, it means, possibly stats are not up to date. So, I'm not sure how such stats are used. Pls share if you have some idea.

Reply:
As long as the writeId associated with the stats is valid according to  the given query's valid writeId list, the stats will be used if they are marked valid. Usually when a writeId advances, the stats will be marked invalid if the operation advancing the writeId renders stats inaccurate. In case of migration, even though writeid advances, the operation may not necessarily render stats inaccurate. In that case, even if the writeId associated with the stats is behind the latest allocated one the stats will be useful as long as 1. the writeId appears valid according to the query's writeId list and 2. stats themselves are marked valid.